raster-foundry/raster-foundry
# Send emails when people request to join groups or are invited

## Overview

This PR adds email notification to creation of user group roles when the creation of
the role results in a `REQUESTED` or `INVITED` status.

### Checklist


### Notes

There's still some cleanup to do to get this to mergeable --

- ~I need tests obviously (there are bombs to make scapegoat angry so I can't forget them)~
- I want to kill the repetitive pattern match [in `Notify.scala`](https://github.com/raster-foundry/raster-foundry/blob/e50531eec7a82bfea243477882d0c2f0811ecc08/app-backend/db/src/main/scala/com/azavea/rf/database/notification/Notify.scala#L38-L58)
- There's no copy yet for the messages, and the HTML messages aren't created yet
- ~Changing roles is broken :man_facepalming: <- this will be the trigger for removing `wip`~

But the basics work.

## Testing Instructions

 * Add email settings to your root platform based on #3578 (they don't need to work, they just need to exist I think)
 * make the dev user a platform admin
 * Create a team with the admin user
 * Note its id and what organization it's under
 * Make the dev user an admin of that team
 * Send a POST request to `/api/platforms/<id>/organization/<id>/teams/<id>/members` to add a different user to that team
 * Watch your logs and note the email
 * Delete the user group role you just created
 * Sign in as a different user
 * Get the token
 * Send the same request
 * Watch your api-server logs and note the email

#### Verification that changing roles works ####

 * log in as the admin user again
 * add a user to the team you created
 * manually set that status to `APPROVED` in the database (or get a token for that user and set it by sending another POST request, but setting it manually is faster)
 * as the admin user, toggle that users group role a few times for fun, ending on ADMIN
 * log in as the user you made an admin
 * confirm that you can toggle other users' roles within that team
 * toggle the team admin user's role (the one you're logged in as) to `MEMBER`
 * try to toggle it back -- note that it fails to change anything

Closes #3552
